### PR TITLE
Python 2/3 compatibility

### PR DESCRIPTION
**Pull request by chadnetzer**
Ongoing work to support both Python 2.6+ and 3 simultaneously.  The print_function import implies minimum Python version of 2.6.  Might be worth officially only supporting 2.7+ if we eventually merge this (as testing on pre-2.7 is increasingly difficult).

### CODE REVIEW

**Comment by chadnetzer** [comment]
@akaihola and reviewers, I suggest prioritizing review and approval of this PR, as it touches many areas of the code (due to all the print statement changes), and thus makes submitting other PRs that aren't based upon it difficult (they will require a lot of extra merging later, if not first based off this merge).

The updates themselves should be pretty uncontroversial, imo, as the "2to3" tool basically gives the same results (though all squashed together).

Note that I specifically avoided importing the Python 2.6+ print function, as the code can currently still remain compatible with earlier versions of Python 2, while simultaneously supporting Python 3.  I think this compatibility is a worthy and important goal for the time being, with minimal downside.

I'm willing to do the work of updating other current PRs which will gain extra merge conflicts due to these changes.

**Comment by chadnetzer** [comment]
Moving this back to a WIP PR, and postponing.  There are enough new PRs, based on current master, that it'll be a lot simpler to update and merge this branch after the dust has settled with the other PRs.  This branch is easy to update as it's mostly syntax changes, whereas merging other PRs after this one will involve re-checking the logic after the merge.

**Comment by chadnetzer** [comment]
Reopening in order to have a published and ready to go python2/3 compatible version based on current master.

Adding py3 compatibility was the original itch I intended to scratch when forking this repo, and this is a mergeable PR that fulfills that need.

File: hardlink.py [hunk 1]
```diff
@@ -114,30 +114,32 @@ def eligible_for_hardlink(st1,        # first file's status
...
-        print st2
-        print "Already hardlinked: %s" % (not is_already_hardlinked(st1, st2))
-        print "Modes:", st1.st_mode, st2.st_mode
-        print "UIDs:", st1.st_uid, st2.st_uid
-        print "GIDs:", st1.st_gid, st2.st_gid
-        print "SIZE:", st1.st_size, st2.st_size
-        print "MTIME:", st1.st_mtime, st2.st_mtime
-        print "Ignore date:", options.notimestamp
-        print "Device:", st1.st_dev, st2.st_dev
+        print("\n***\n", st1)
```
Comment (akaihola) [inline hardlink.py]: This will change output on Python 2 from

    <empty line>
    ***
    <string value of st1>

to

    ('\n***\n', <repr of foo>)

File: hardlink.py [hunk 2]
```diff
@@ -154,21 +156,24 @@ def hardlink_files(sourcefile, destfile, stat_info, options):
     try:
         if not options.dryrun:
             os.rename(destfile, temp_name)
-    except OSError as error:
-        print "Failed to rename: %s to %s: %s" % (destfile, temp_name, error)
+    except OSError:
```
Comment (akaihola) [inline hardlink.py]: Why is this change needed? I tried on Python 2.7, 3.5 and 3.6, and in all of those using `except ... as ...` works fine, and the string representation of the exception object looks good.

There's just a slight difference in `str(error)`:

    '[Errno 2] No such file or directory'  # Python 2.7
    "[Errno 2] No such file or directory: 'foo' -> 'bar'"  # Python 3.5 & 3.6

File: hardlink.py (hunk 1 above)
Comment (chadnetzer) [reply hardlink.py]: This whole section is disabled by "if None", so I vote to delete it.  But if kept, it can and should be fixed to work equivalently for both Py 2 and 3.

File: hardlink.py (hunk 2 above)
Comment (chadnetzer) [reply hardlink.py]: I'm targeting Python 2.3 and above (for now), as it still the norm on RHEL 5 and others.  If you want to target Python 2.7+, that's fine with me too.  Given all the current churn, it might be nice to do 1 additional release with older Python 2 support, before committing to newer (Python 2.7 or even 3 only).

Comment (akaihola) [reply hardlink.py]: Ah, so these changes not only provide Python 3 compatibility but also extend supported Python versions all the way back to Python 2.3. That's sounds reasonable.

**Comment by chadnetzer** [comment]
> '%s' % value implies str(value)

Surprisingly there's a subtlety here: on Python 3 "%s" % value fails with the stat_result object as a value because it sees it as a tuple-like (with 10-ish elements) which it tries to unpack for the string formatting.  You discuss this scenario in your follow up comment.  Using str() "fixes" this, although '%s' % (value,) also works for both Py2 and Py3.
